File: ch4/sssp/bellman_ford_moore.ml
open Scanf
open Printf

let solve case =
  let n, m, s, t = scanf "%d %d %d %d\n" (fun x y z t -> (x, y, z, t)) in
  let adj = Array.make n [] in
  for i = 1 to m do
    let u, v, w = scanf "%d %d %d\n" (fun x y z -> (x, y, z)) in
    adj.(u) <- (v, w) :: adj.(u);
    adj.(v) <- (u, w) :: adj.(v)
  done;
  (* SPFA algorithm *)
  let dist = Array.make n max_int in
  let in_queue = Array.make n false in
  let q = Queue.create () in
  dist.(s) <- 0;
  Queue.add s q;
  in_queue.(s) <- true;
  while not (Queue.is_empty q) do
    let u = Queue.pop q in
    in_queue.(u) <- false;
    if dist.(u) = max_int then () else
    adj.(u) |> List.iter (fun (v, w) ->
      if dist.(u) + w < dist.(v) then begin
        dist.(v) <- dist.(u) + w;
        if not in_queue.(v) then begin
          Queue.add v q;
          in_queue.(v) <- true
        end
      end
    )
  done;
  printf "Case #%d: " case;
  if dist.(t) = max_int then printf "unreachable\n"
  else printf "%d\n" dist.(t)

let () =
  let tc = scanf "%d\n" (fun x -> x) in
  for i = 1 to tc do solve i done

File: ch4/sssp/bellman_ford_moore.py
from collections import deque

def main():
    INF = int(1e9)

    f = open("bellman_ford_in.txt", "r")

    V, E, s = map(int, f.readline().split(" "))
    AL = [[] for u in range(V)]
    for _ in range(E):
        u, v, w = map(int, f.readline().split(" "))
        AL[u].append((v, w))

    # SPFA from source S
    # initially, only source vertex s has dist[s] = 0 and in the queue
    dist = [INF for u in range(V)]
    dist[s] = 0
    q = deque()
    q.append(s)
    in_queue = [0 for u in range(V)]
    in_queue[s] = 1
    while (len(q) > 0):
        u = q.popleft()                          # pop from queue
        in_queue[u] = 0
        for v, w in AL[u]:
            if (dist[u]+w >= dist[v]): continue  # not improving, skip
            dist[v] = dist[u]+w                  # relax operation
            if not in_queue[v]:                  # add to the queue
                q.append(v)                      # only if v is not
                in_queue[v] = 1                  # already in the queue

    for u in range(V):
        print("SSSP({}, {}) = {}".format(s, u, dist[u]))

main()
